presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -979,10 +981,30 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
                     throw new SemanticException(INVALID_PROCEDURE_ARGUMENTS, initialValueArg, "REDUCE_AGG only supports non-NULL literal as the initial value", initialValueArg);
                 }
             }
+
             Type type = functionAndTypeManager.getType(functionMetadata.getReturnType());
+
+            if (type instanceof MapType) {
+                Type keyType = ((MapType) type).getKeyType();
+                if (TypeUtils.isApproximateNumericType(keyType)) {
+                    String warningMessage = createWarningMessage(node, "This query uses double as map key which can have non-deterministic results.");
+                    warningCollector.add(new PrestoWarning(SEMANTIC_WARNING, warningMessage));
```

Comment:
Make the message actionable.

Map keys with real/double type can be non-deterministic. Please use decimal type instesad